### PR TITLE
Let field tests work with 256-bit curves like secp256k1

### PR DESCRIPTION
**Pull request by weikengchen**
## Description

When implementing secp256k1, we find that the field tests do not work well with curves whose modulus bit numbers are 64 * K.

This PR fixes so.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

N/A:
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`

### CODE REVIEW

File: serialize/src/lib.rs [hunk 1]
```diff
@@ -204,15 +204,15 @@ impl<T: CanonicalSerialize> CanonicalSerializeHashExt for T {}
 
 #[inline]
 pub fn buffer_bit_byte_size(modulus_bits: usize) -> (usize, usize) {
-    let byte_size = buffer_byte_size(modulus_bits);
+    let byte_size = buffer_byte_size(modulus_bits + 1);
```
Comment (weikengchen) [inline serialize/src/lib.rs:207]:
```suggestion
    let byte_size = buffer_byte_size(modulus_bits);
```

File: test-templates/src/fields.rs [hunk 2]
```diff
@@ -54,7 +54,7 @@ macro_rules! __test_field {
                     let buf_size = <$field>::zero().serialized_size(compress);
 
                     let buffer_size =
-                        buffer_bit_byte_size(<$field as Field>::BasePrimeField::MODULUS_BIT_SIZE as usize).1 *
+                        buffer_bit_byte_size(<$field as Field>::BasePrimeField::MODULUS_BIT_SIZE as usize + 1).1 *
```
Comment (weikengchen) [inline test-templates/src/fields.rs:57]:
```suggestion
                        buffer_bit_byte_size(<$field as Field>::BasePrimeField::MODULUS_BIT_SIZE as usize).1 *
```

File: test-templates/src/fields.rs [hunk 3]
```diff
@@ -395,7 +403,11 @@ macro_rules! __test_field {
             assert_eq!(BigUint::from(<$field>::MODULUS_MINUS_ONE_DIV_TWO), &modulus_minus_one / 2u32);
             assert_eq!(<$field>::MODULUS_BIT_SIZE as u64, modulus.bits());
             if let Some(SqrtPrecomputation::Case3Mod4 { modulus_plus_one_div_four }) = <$field>::SQRT_PRECOMP {
-                assert_eq!(modulus_plus_one_div_four, &((&modulus + 1u8) / 4u8).to_u64_digits());
+                let mut expected = ((&modulus + 1u8) / 4u8).to_u64_digits();
+                if <$field>::MODULUS_BIT_SIZE as u64 % 64 == 0 {
```
Comment (Pratyush) [inline test-templates/src/fields.rs:407]: Why is this necessary? Modulus_plus_one_div_four has always 255 bits at most, and so should fit in 4 limbs

File: serialize/src/lib.rs [hunk 4]
```diff
@@ -212,7 +212,7 @@ pub fn buffer_bit_byte_size(modulus_bits: usize) -> (usize, usize) {
 /// into the number of bytes required to represent it.
 #[inline]
 pub const fn buffer_byte_size(modulus_bits: usize) -> usize {
-    (modulus_bits + 7) / 8
+    (modulus_bits + 1 + 7) / 8
```
Comment (Pratyush) [inline serialize/src/lib.rs:215]: This isn’t necessarily correct; it adds a byte to every representation, no? Can we add a test to ensure that it maintains the old behaviour for smaller moduli?

and is the reason the test is failing because you’re trying to use the existing arithmetic infrastructure for 256 bit moduli?

Comment (Pratyush) [reply serialize/src/lib.rs:215]: If you’re trying to use the existing infrastructure for secp256k1, I think a better idea would be to port over the algorithms from here: https://github.com/pornin/crrl/blob/3a3e625d784b4234d53829771fc3e18abab16913/src/backend/w64/modint.rs#L141

that way we introduce the extra limb only when necessary.

**Comment by weikengchen** [comment]
I will come back and cross-check on this.

File: serialize/src/lib.rs (hunk 4 above)
Comment (weikengchen) [reply serialize/src/lib.rs:221]: Let me take a look at this new algorithm---it would be a cleaner fix and is also for the future.

**Comment by weikengchen** [comment]
This PR is no longer the solution---instead of allowing an additional limb, we now generalize the code into supporting full limbs. Closed.
